server/src/main/java/org/elasticsearch/health/HealthService.java

```diff
@@ -80,30 +98,96 @@ public List<HealthIndicatorResult> getHealth(@Nullable String indicatorName, boo
         Stream<HealthIndicatorService> filteredIndicators = healthIndicatorServices.stream()
             .filter(service -> indicatorName == null || service.name().equals(indicatorName));
 
-        Stream<HealthIndicatorResult> filteredIndicatorResults;
         if (clusterHealthIsObtainable) {
-            // Calculate remaining indicators
-            filteredIndicatorResults = filteredIndicators.map(service -> service.calculate(explain));
+            if (HealthNode.isEnabled()) {
+                try {
+                    client.execute(FetchHealthInfoCacheAction.INSTANCE, new FetchHealthInfoCacheAction.Request(), new ActionListener<>() {
+                        @Override
+                        public void onResponse(FetchHealthInfoCacheAction.Response response) {
+                            HealthInfo healthInfo = response.getHealthInfo();
+                            combineResultsAndNotifyListener(
+                                listener,
+                                indicatorName,
+                                preflightResults,
+                                filteredIndicators.map(service -> service.calculate(explain, healthInfo)).toList()
+                            );
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            getHealthNoHealthInfo(listener, indicatorName, preflightResults, filteredIndicators.toList(), explain);
+                        }
+                    });
+                } catch (NodeNotConnectedException | HealthNodeNotDiscoveredException e) {
```

Comment:
I believe the async calls don't throw these exceptions right? They're propagated to the listener via `onFailure`

Do we need special handling there?

Reply:
I believe that this happens in `client.execute` before it does the async bit but i'll double-check.